### PR TITLE
Us - Event_API

### PR DESCRIPTION
**Pull request by MareikeFla**
It is done!
Events from [the event calender of the city of cologne](https://www.govdata.de/web/guest/suchen/-/details/veranstaltungen-der-stadt-koln) are fetched and included in our event database on a daily base. 
I have deviated from the US in some points because I was learning while doing it. 
I really thought about how to break this PR into smaller pieces but I haven't found a way that makes sense - I am sorry... but have fun!

### CODE REVIEW

File: lib/daylieFetchHelper/functions.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    method: "DELETE",
+  });
+}
+
+export function findUniqueEventsInFirstArray(eventArray1, eventArray2) {
+  const fetchedIdInArray2 = new Set(
+    eventArray2.map((event) => event.fetchedId)
+  );
+  return eventArray1.filter((event) => !fetchedIdInArray2.has(event.fetchedId));
+}
```
Comment (MerleMak) [inline lib/daylieFetchHelper/functions.js:148]: I think this file needs a better name :) also maybe there is a better spot for this file than the dailyFetchHelper, as this a very specific use case. 
Let me know if you are planning on restructuring the functions anyway, but you can do that to a later point in a refactoring ticket :)

**Review by MerleMak — COMMENTED** [review]
Wow what a task! You did an amazing job! 
I left you some comments for you to consider. 
Really really nice, keep it up Mareike

File: lib/daylieFetchHelper/formatingData.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
+import { adminID } from "./variables";
+
+function extractIDFromUrl(url) {
+  const linkPattern =
+    /^https?:\/\/www\.stadt-koeln\.de\/leben-in-koeln\/veranstaltungen\/daten\/(.+?)\/index\.html/;
+  const match = url.match(linkPattern);
+  return match ? match[1] : null;
```
[the remark this answers is not in the document]
Comment (MareikeFla) [reply lib/daylieFetchHelper/formatingData.js:7]: Unfortunately the API does not provide an event ID, but the id is part of the event link. The url looks like this:
https://www.stadt-koeln.de/leben-in-koeln/veranstaltungen/daten/34893/index.html

match[0] is the first part of the pattern "http://www.stadt-koeln.de/leben-in-koeln/veranstaltungen/daten/"
match[1] is the event id "34893"
match[2] is "/index.html"

File: lib/daylieFetchHelper/formatingData.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  if (string.length <= maxLength) {
+    return string;
+  } else {
+    const lastSpaceIndex = string.lastIndexOf(" ", schortendMaxLength);
+    return string.slice(0, lastSpaceIndex) + "...";
+  }
+}
+
+function expandDescription(oepnv, uhrzeit, description) {
+  let newDescription = description;
```
[the remark this answers is not in the document]
Comment (MareikeFla) [reply lib/daylieFetchHelper/formatingData.js:22]: Yes! sounds good

File: lib/daylieFetchHelper/formatingData.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    newDescription = description + "\n\nInformationen zur Uhrzeit: " + uhrzeit;
+  }
+  return newDescription;
+}
+
+function parseTimeString(input) {
+  // This regular expression is used to match time ranges in a specific format.
+  // It is useful for parsing strings that denote time intervals, such as '10:30 Uhr bis 14:45 Uhr' or '9 bis 12', verifying their format and extracting the times.
+  const timePattern =
+    /^(\d{1,2})(?::(\d{2}))?\s*(Uhr)?\s*bis\s*(\d{1,2})(?::(\d{2}))?(\s*Uhr)?$/;
```
[the remark this answers is not in the document]
Comment (MareikeFla) [reply lib/daylieFetchHelper/formatingData.js:37]: It took me two ours to figure out a version that worked if the time was a format like "12:00 bis 14:00" 😂
After that chatGPT helped me out, the hardest part was, that there are so many different time formats coming from that api....

File: lib/daylieFetchHelper/variables.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+export const adminID = "661e7bfaa3b5c569e6342d1a";
```
[the remark this answers is not in the document]
Comment (MareikeFla) [reply lib/daylieFetchHelper/variables.js:1]: That is kind of a good point 😅
At the moment it is just a user account I created to provide a userID to the createdBy key, but nevertheless there should be no visible userIds - will fix

File: lib/daylieFetchHelper/formatingData.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      links: [
+        {
+          url: link || "",
+          linkDescription: link ? "Mehr Informationen" : "",
+        },
+      ],
+    };
+  });
+  return events;
+}
```
[the remark this answers is not in the document]
Comment (MareikeFla) [reply lib/daylieFetchHelper/formatingData.js:168]: 😭😂

File: lib/daylieFetchHelper/functions.js (hunk 1 above)
Comment (MareikeFla) [reply lib/daylieFetchHelper/functions.js]: Yes, you are right. The structure will be refactored in the next sprint :)

File: lib/daylieFetchHelper/formatingData.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (MerleMak) [reply lib/daylieFetchHelper/formatingData.js]: interesting!! thank you for the explanation :)

File: lib/daylieFetchHelper/formatingData.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (MerleMak) [reply lib/daylieFetchHelper/formatingData.js]: crazy!! and very impressive!

**Review by MerleMak — APPROVED** [review]
great!!

**Pull request merged**